File: lp5-datasets/05 Mini Huffman Encode/backup(if nothing works)/mini_huffman_encode.py
# Mini Project: Implement Huffman Encoding on GPU

import heapq
from collections import Counter
from multiprocessing import Pool
import time


# Node Class
class Node:
    def __init__(self, char, freq):
        self.char = char
        self.freq = freq
        self.left = None
        self.right = None

    def __lt__(self, other):
        return self.freq < other.freq


# Build Huffman Tree
def build_tree(text):
    freq = Counter(text)
    heap = [Node(char, freq) for char, freq in freq.items()]
    heapq.heapify(heap)

    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)

        merged = Node(None, left.freq + right.freq)
        merged.left = left
        merged.right = right

        heapq.heappush(heap, merged)

    return heap[0]


# Generate Codes
def generate_codes(node, prefix="", codebook={}):
    if node is None:
        return

    if node.char is not None:
        codebook[node.char] = prefix

    generate_codes(node.left, prefix + "0", codebook)
    generate_codes(node.right, prefix + "1", codebook)

    return codebook


# Encode Chunk (Parallel)
def encode_chunk(args):
    chunk, codebook = args
    return "".join([codebook[ch] for ch in chunk])


# Parallel Encoding
def parallel_encode(text, codebook, num_workers=4):
    chunk_size = len(text) // num_workers
    chunks = [text[i : i + chunk_size] for i in range(0, len(text), chunk_size)]

    with Pool(num_workers) as p:
        result = p.map(encode_chunk, [(chunk, codebook) for chunk in chunks])

    return "".join(result)


# Sequential Encoding
def sequential_encode(text, codebook):
    return "".join([codebook[ch] for ch in text])


# MAIN
if __name__ == "__main__":
    text = "this is a huffman encoding test " * 10000

    # Build tree
    root = build_tree(text)
    codebook = generate_codes(root)

    # Sequential
    start = time.time()
    seq_encoded = sequential_encode(text, codebook)
    print("Sequential Time:", time.time() - start)

    # Parallel
    start = time.time()
    par_encoded = parallel_encode(text, codebook)
    print("Parallel Time:", time.time() - start)

    print("Encoding successful:", seq_encoded == par_encoded)

File: lp5-datasets/05 Mini Huffman Encode/backup(if nothing works)/mini_huffman_encode.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>
#include <string>
#include <omp.h>

using namespace std;

// Node structure
struct Node
{
    char ch;
    int freq;
    Node *left, *right;

    Node(char c, int f)
    {
        ch = c;
        freq = f;
        left = right = nullptr;
    }
};

// Comparator for priority queue
struct compare
{
    bool operator()(Node *l, Node *r)
    {
        return l->freq > r->freq;
    }
};

// Build Huffman Tree (SEQUENTIAL)
Node *buildTree(const string &text)
{
    unordered_map<char, int> freq;

    for (char ch : text)
        freq[ch]++;

    priority_queue<Node *, vector<Node *>, compare> pq;

    for (auto pair : freq)
        pq.push(new Node(pair.first, pair.second));

    while (pq.size() > 1)
    {
        Node *left = pq.top();
        pq.pop();
        Node *right = pq.top();
        pq.pop();

        Node *merged = new Node('\0', left->freq + right->freq);
        merged->left = left;
        merged->right = right;

        pq.push(merged);
    }

    return pq.top();
}

// Generate Huffman Codes (SEQUENTIAL)
void generateCodes(Node *root, string code,
                   unordered_map<char, string> &codebook)
{
    if (!root)
        return;

    if (root->ch != '\0')
        codebook[root->ch] = code;

    generateCodes(root->left, code + "0", codebook);
    generateCodes(root->right, code + "1", codebook);
}

// Parallel Encoding (OPENMP)
string parallelEncode(const string &text,
                      unordered_map<char, string> &codebook)
{

    vector<string> encoded(text.size());

#pragma omp parallel for
    for (int i = 0; i < text.size(); i++)
    {
        encoded[i] = codebook[text[i]];
    }

    // Combine results
    string result = "";
    for (auto &s : encoded)
        result += s;

    return result;
}

// Sequential Encoding
string sequentialEncode(const string &text,
                        unordered_map<char, string> &codebook)
{
    string result = "";
    for (char ch : text)
        result += codebook[ch];

    return result;
}

// MAIN
int main()
{
    string text = "this is a huffman encoding test ";

    // Increase workload
    string largeText = "";
    for (int i = 0; i < 10000; i++)
        largeText += text;

    // Build tree
    Node *root = buildTree(largeText);

    unordered_map<char, string> codebook;
    generateCodes(root, "", codebook);

    // Sequential
    double start = omp_get_wtime();
    string seq = sequentialEncode(largeText, codebook);
    double end = omp_get_wtime();
    cout << "Sequential Time: " << (end - start) << " sec\n";

    // Parallel
    start = omp_get_wtime();
    string par = parallelEncode(largeText, codebook);
    end = omp_get_wtime();
    cout << "Parallel Time: " << (end - start) << " sec\n";

    cout << "Encoding correct: " << (seq == par) << endl;

    return 0;
}